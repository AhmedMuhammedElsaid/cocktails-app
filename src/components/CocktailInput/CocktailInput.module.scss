@import "../../scss/variables.scss";
.Container {
  margin-bottom: 10px;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding: 10px;

  &__Head {
    display: flex;
    align-items: center;
    h2 {
      margin-bottom: 5px;
    }
    p {
      color: $lightGrey;
    }
  }
  &__Input__Icon {
    position: relative;
    border: 1px solid $lightGrey;
    background-color: $mainWhite;
    margin-bottom: 5px;
    padding: 10px 20px;
    margin-top: 5px;
    input {
      color: $darkGrey;
      border: none;
      outline: none;
      font-size: 14px;
      min-width: 250px;
      width: 300px;
      &::placeholder {
        color: $darkGrey;
        text-transform: capitalize;
      }
      margin-left: 20px;
      @media (max-width: 768px) {
        margin-bottom: 20px;
      }
      @media (max-width: 768px) {
        width: 100% !important;
      }
    }
    img {
      position: absolute;
      left: 0.8em;
      top: 1.2em;
    }
  }
}
