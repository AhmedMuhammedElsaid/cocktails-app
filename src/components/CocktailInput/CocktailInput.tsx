import React, { FunctionComponent } from "react";
import styles from "./CocktailInput.module.scss";
import SearchImg from "../../assets/images/searchIcon.png";
interface Props {
  value?: string;
  handleChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
}
const CocktailInput: FunctionComponent<Props> = ({ value, handleChange }) => {
  return (
    <div className={styles.Container}>
      <div className={styles.Container__Head}>
        <div>
          <h2>Cocktail Searcher</h2>
          <p>
            browse through all cocktails based on initial letter <br />
            cocktails with the initial letter will show up immediately{" "}
          </p>
        </div>
      </div>
      <div className={styles.Container__Input__Icon}>
        <input
          value={value}
          onChange={handleChange}
          name="name"
          data-testid="cocktail-input-testid"
          autoComplete="false"
          placeholder="search by initial cocktail letter here...."
          type="search"
        />
        <img src={SearchImg} alt="search icon" />
      </div>
    </div>
  );
};

export default CocktailInput;
