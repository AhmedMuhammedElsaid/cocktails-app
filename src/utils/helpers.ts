/*
 @params object
 return object with no null ,undefined , empty string "cleaned object"
*/
export function cleanObject(object: { [x: string]: any }) {
  if (!object) return {};
  if (typeof object === "object") {
    for (let propName in object) {
      if (
        object[propName] === null ||
        object[propName] === undefined ||
        object[propName]?.length === 0
      ) {
        delete object[propName];
      }
    }
  }
  return object;
}
