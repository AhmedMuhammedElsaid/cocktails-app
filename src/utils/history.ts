import { createBrowserHistory } from "history";

const history = createBrowserHistory();
// method to redirect the user to error page in case any error happens while fetching the data from the server
export const redirectToErrorPage = () => {
  history.push(`/pagenotfound`);
  history.go(0);
};
export default history;
