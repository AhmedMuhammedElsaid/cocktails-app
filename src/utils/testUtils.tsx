import { render, RenderResult } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const customRender = (component: JSX.Element): RenderResult => {
  return {
    ...render(<Router>{component}</Router>),
  };
};
