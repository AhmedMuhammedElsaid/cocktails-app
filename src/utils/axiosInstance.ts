import axios from "axios";
import { redirectToErrorPage } from './history';
import { toast } from "react-toastify";

const axiosInstance = axios.create({
  // setting the default uri for this instance
  baseURL: process.env.REACT_APP_BASE_URI,
});

// setting language  configuration on every single request we send to the BE
axiosInstance.interceptors.request.use(
  (request: any) => {
    return {
      ...request,
      headers: {
        ...request.headers,
        // setting the header lang german "de"
        "Accept-Language": "de",
      },
    };
  },
  (error) => {
    return Promise.reject(error);
  }
);

axiosInstance.interceptors.response.use(
  (response) => {
    return response.data;
  },
  (error) => {
    switch (error?.response?.status) {
      case 400:
        return {
          data: null,
          hasError: true,
          error: [error.response.data],
        };
      case 401:
        toast.error("Unauthorized Call Access");
        break;
      case 404:
        toast.error("404 Not Found");
        break;
      case 500:
        toast.error("Server Is Down");
        break;
    }
    redirectToErrorPage()
    return Promise.reject(error);
  }
);
export default axiosInstance;
