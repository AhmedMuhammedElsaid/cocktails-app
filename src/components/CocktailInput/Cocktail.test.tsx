import CocktailInput from "./CocktailInput";
import { fireEvent } from "@testing-library/react";
import { customRender } from "utils/testUtils";
import "@testing-library/jest-dom";

const cocktailInput = "cocktail-input-testid";
describe("Test Cocktail Input Component", () => {
  it("should render input and simulate the user typing action", () => {
    const { asFragment, getByTestId } = customRender(
      <CocktailInput handleChange={jest.fn()} />
    );
    fireEvent.change(getByTestId(cocktailInput), {
      target: { value: "jon" },
    });
    expect(getByTestId(cocktailInput)).toHaveValue("jon");
    expect(asFragment()).toMatchSnapshot();
  });
});
