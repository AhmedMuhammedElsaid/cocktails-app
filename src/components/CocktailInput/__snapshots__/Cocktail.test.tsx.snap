// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Cocktail Input Component should render input and simulate the user typing action 1`] = `
<DocumentFragment>
  <div
    class="Container"
  >
    <div
      class="Container__Head"
    >
      <div>
        <h2>
          Cocktail Searcher
        </h2>
        <p>
          browse through all cocktails based on initial letter 
          <br />
          cocktails with the initial letter will show up immediately 
        </p>
      </div>
    </div>
    <div
      class="Container__Input__Icon"
    >
      <input
        autocomplete="false"
        data-testid="cocktail-input-testid"
        name="name"
        placeholder="search by initial cocktail letter here...."
        type="search"
        value=""
      />
      <img
        alt="search icon"
        src="searchIcon.png"
      />
    </div>
  </div>
</DocumentFragment>
`;
